core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1441,6 +1441,18 @@ class DAGScheduler(
             failedStages += failedStage
             failedStages += mapStage
             if (noResubmitEnqueued) {
+              if (!mapStage.rdd.isIdempotent) {
+                // The map stage is not idempotent, we have to rerun all the tasks for the
+                // failed stage to get expected result.
+                failedStage match {
+                  case s: ShuffleMapStage =>
```

Comment:
Like we discussed, we shall also retry the partially finished succeeding stages.

Reply:
We may also have to update the logic in `removeExecutorAndUnregisterOutputs`.